File: ios/Package/Sources/LensPilotCore/CreativeInterpretation.swift
import Foundation

public enum CreativeInterpretationAdapterError: Error, Equatable, Sendable {
    case blockedByHealthGate([CreativeInterpretationProviderHealthGate.DeniedReason])
    case emptyProviderOutput
    case unsafeProviderResponse
}

public struct CreativeInterpretationProviderHealthGate: Codable, Equatable, Sendable {
    public let canRunProvider: Bool
    public let deniedReasons: [DeniedReason]
    public let providerHealthStatus: OnlineInspirationHealthSnapshot.Status?
    public let publicReferenceCount: Int
    public let payloadAudit: CreativeInterpretationPayloadAudit
    public let privacy: Privacy

    public init(
        canRunProvider: Bool,
        deniedReasons: [DeniedReason],
        providerHealthStatus: OnlineInspirationHealthSnapshot.Status?,
        publicReferenceCount: Int,
        payloadAudit: CreativeInterpretationPayloadAudit,
        privacy: Privacy
    ) {
        let uniqueDeniedReasons = Array(Set(deniedReasons)).sorted { $0.rawValue < $1.rawValue }

        self.canRunProvider = canRunProvider && uniqueDeniedReasons.isEmpty
        self.deniedReasons = uniqueDeniedReasons
        self.providerHealthStatus = providerHealthStatus
        self.publicReferenceCount = max(0, publicReferenceCount)
        self.payloadAudit = payloadAudit
        self.privacy = privacy
    }

    public static func make(
        for request: CreativeInterpretationRequest,
        healthSnapshot: OnlineInspirationHealthSnapshot?
    ) -> CreativeInterpretationProviderHealthGate {
        let privacy = Privacy(requestPrivacy: request.privacy, healthSnapshotPrivacy: healthSnapshot?.privacy)
        var deniedReasons: [DeniedReason] = []

        if !request.payloadAudit.safeToSend {
            deniedReasons.append(.unsafeRequestPayload)
        }

        guard let healthSnapshot else {
            deniedReasons.append(.missingProviderHealth)
            return CreativeInterpretationProviderHealthGate(
                canRunProvider: false,
                deniedReasons: deniedReasons,
                providerHealthStatus: nil,
                publicReferenceCount: 0,
                payloadAudit: request.payloadAudit,
                privacy: privacy
            )
        }

        if !privacy.isSafeForSinglePhoneProvider
            || !healthSnapshot.providers.allSatisfy({ isSafeProviderPrivacy($0.privacy) }) {
            deniedReasons.append(.unsafeProviderHealth)
        }

        switch healthSnapshot.status {
        case .available, .degraded:
            break
        case .empty:
            deniedReasons.append(.noPublicReferences)
        case .failed:
            deniedReasons.append(.providerUnavailable)
        }

        if healthSnapshot.totalResultCount <= 0 {
            deniedReasons.append(.noPublicReferences)
        }

        return CreativeInterpretationProviderHealthGate(
            canRunProvider: true,
            deniedReasons: deniedReasons,
            providerHealthStatus: healthSnapshot.status,
            publicReferenceCount: healthSnapshot.totalResultCount,
            payloadAudit: request.payloadAudit,
            privacy: privacy
        )
    }

    private static func isSafeProviderPrivacy(_ privacy: OnlineInspirationProviderHealth.Privacy) -> Bool {
        privacy.publicSourceOnly
            && privacy.derivedFromPromptOnly
            && !privacy.storesRawPhoto
            && !privacy.uploadsLiveCameraFrame
            && !privacy.sendsIdentityData
            && !privacy.sendsPreciseLocation
    }
}

public extension CreativeInterpretationProviderHealthGate {
    enum DeniedReason: String, Codable, Equatable, Sendable, Hashable {
        case unsafeRequestPayload = "unsafe_request_payload"
        case missingProviderHealth = "missing_provider_health"
        case unsafeProviderHealth = "unsafe_provider_health"
        case providerUnavailable = "provider_unavailable"
        case noPublicReferences = "no_public_references"
    }

    struct Privacy: Codable, Equatable, Sendable {
        public let singlePhoneOnly: Bool
        public let requiresUserConsent: Bool
        public let sendsRawCameraFrame: Bool
        public let sendsPrivatePhoto: Bool
        public let sendsIdentityData: Bool
        public let sendsPreciseLocation: Bool
        public let sendsRawLearningEvents: Bool

        public init(
            singlePhoneOnly: Bool = true,
            requiresUserConsent: Bool = true,
            sendsRawCameraFrame: Bool = false,
            sendsPrivatePhoto: Bool = false,
            sendsIdentityData: Bool = false,
            sendsPreciseLocation: Bool = false,
            sendsRawLearningEvents: Bool = false
        ) {
            self.singlePhoneOnly = singlePhoneOnly
            self.requiresUserConsent = requiresUserConsent
            self.sendsRawCameraFrame = sendsRawCameraFrame
            self.sendsPrivatePhoto = sendsPrivatePhoto
            self.sendsIdentityData = sendsIdentityData
            self.sendsPreciseLocation = sendsPreciseLocation
            self.sendsRawLearningEvents = sendsRawLearningEvents
        }

        public init(
            requestPrivacy: CreativeInterpretationPlan.Privacy,
            healthSnapshotPrivacy: OnlineInspirationHealthSnapshot.Privacy?
        ) {
            self.init(
                singlePhoneOnly: requestPrivacy.singlePhoneOnly && (healthSnapshotPrivacy?.singlePhoneOnly ?? true),
                requiresUserConsent: requestPrivacy.requiresUserConsent && (healthSnapshotPrivacy?.requiresUserConsent ?? true),
                sendsRawCameraFrame: requestPrivacy.sendsRawCameraFrame || (healthSnapshotPrivacy?.sendsRawCameraFrame ?? false),
                sendsPrivatePhoto: requestPrivacy.sendsPrivatePhoto || (healthSnapshotPrivacy?.sendsPrivatePhoto ?? false),
                sendsIdentityData: requestPrivacy.sendsIdentityData || (healthSnapshotPrivacy?.sendsIdentityData ?? false),
                sendsPreciseLocation: requestPrivacy.sendsPreciseLocation || (healthSnapshotPrivacy?.sendsPreciseLocation ?? false),
                sendsRawLearningEvents: requestPrivacy.sendsRawLearningEvents
            )
        }

        public var isSafeForSinglePhoneProvider: Bool {
            singlePhoneOnly
                && requiresUserConsent
                && !sendsRawCameraFrame
                && !sendsPrivatePhoto
                && !sendsIdentityData
                && !sendsPreciseLocation
                && !sendsRawLearningEvents
        }
    }
}

public struct CreativeInterpretationProviderResult: Codable, Equatable, Sendable {
    public let headline: String
    public let guidance: [String]

    public init(headline: String, guidance: [String]) {
        self.headline = Self.cleanedText(headline, maxLength: 96)
        self.guidance = guidance
            .map { Self.cleanedText($0, maxLength: 180) }
            .filter { !$0.isEmpty }
            .prefix(4)
            .map { $0 }
    }

    private static func cleanedText(_ value: String, maxLength: Int) -> String {
        let collapsed = value
            .split(whereSeparator: \.isWhitespace)
            .joined(separator: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)

        guard collapsed.count > maxLength else { return collapsed }

        return String(collapsed.prefix(maxLength))
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

public protocol CreativeInterpretationReasoningProvider: Sendable {
    var provider: CreativeInterpretationRequest.Provider { get }
    func interpret(request: CreativeInterpretationRequest) async throws -> CreativeInterpretationProviderResult
}

public struct CreativeInterpretationHeuristicProvider: CreativeInterpretationReasoningProvider {
    public let provider: CreativeInterpretationRequest.Provider

    public init(provider: CreativeInterpretationRequest.Provider = .onlineReasoning) {
        self.provider = provider
    }

    public func interpret(request: CreativeInterpretationRequest) async throws -> CreativeInterpretationProviderResult {
        guard request.payloadAudit.safeToSend else {
            throw CreativeInterpretationError.unsafePlan(request.payloadAudit.deniedReasons)
        }

        let guidance = Self.guidance(from: request)
        guard !guidance.isEmpty else {
            throw CreativeInterpretationAdapterError.emptyProviderOutput
        }

        return CreativeInterpretationProviderResult(
            headline: provider == .onlineReasoning ? "Provider-Ready Creative Brief" : "Local Creative Brief",
            guidance: guidance
        )
    }

    private static func guidance(from request: CreativeInterpretationRequest) -> [String] {
        var remainingWords = request.maxResponseWords
        var guidance: [String] = []
        let briefs = prioritizedBriefs(from: request)

        for brief in briefs where remainingWords > 0 {
            let words = brief.split(whereSeparator: \.isWhitespace)
            guard !words.isEmpty else { continue }

            let wordCount = min(words.count, remainingWords)
            guidance.append(words.prefix(wordCount).joined(separator: " "))
            remainingWords -= wordCount
        }

        return guidance
    }

    private static func prioritizedBriefs(from request: CreativeInterpretationRequest) -> [String] {
        guard let safetyBrief = request.suggestionBriefs.first(where: {
            $0.localizedCaseInsensitiveContains("Capture-Realistic")
        }) else {
            return Array(request.suggestionBriefs.prefix(4))
        }

        let primaryBriefs = request.suggestionBriefs
            .filter { $0 != safetyBrief }
            .prefix(3)
        return Array(primaryBriefs) + [safetyBrief]
    }
}

public struct CreativeInterpretationResponse: Identifiable, Codable, Equatable, Sendable {
    public let id: String
    public let planId: String
    public let provider: CreativeInterpretationRequest.Provider
    public let status: Status
    public let headline: String
    public let guidance: [String]
    public let maxResponseWords: Int
    public let generatedAt: Date
    public let payloadAudit: CreativeInterpretationPayloadAudit
    public let healthGate: CreativeInterpretationProviderHealthGate
    public let privacy: Privacy

    public init(
        planId: String,
        provider: CreativeInterpretationRequest.Provider,
        status: Status = .completed,
        result: CreativeInterpretationProviderResult,
        maxResponseWords: Int,
        generatedAt: Date = Date(),
        payloadAudit: CreativeInterpretationPayloadAudit,
        healthGate: CreativeInterpretationProviderHealthGate,
        privacy: Privacy = Privacy()
    ) {
        self.id = "creative_interpretation_response_\(planId)_\(provider.rawValue)"
        self.planId = planId
        self.provider = provider
        self.status = status
        self.headline = result.headline
        self.guidance = result.guidance
        self.maxResponseWords = min(240, max(40, maxResponseWords))
        self.generatedAt = generatedAt
        self.payloadAudit = payloadAudit
        self.healthGate = healthGate
        self.privacy = privacy
    }
}

public extension CreativeInterpretationResponse {
    enum Status: String, Codable, Equatable, Sendable {
        case completed
    }

    struct Privacy: Codable, Equatable, Sendable {
        public let singlePhoneOnly: Bool
        public let usesAuditedPayload: Bool
        public let usesProviderHealthGate: Bool
        public let storesRawPhoto: Bool
        public let uploadsLiveCameraFrame: Bool
        public let sendsPrivatePhoto: Bool
        public let sendsIdentityData: Bool
        public let sendsPreciseLocation: Bool
        public let sendsRawLearningEvents: Bool
        public let allowsGenerativeImageOutput: Bool

        public init(
            singlePhoneOnly: Bool = true,
            usesAuditedPayload: Bool = true,
            usesProviderHealthGate: Bool = true,
            storesRawPhoto: Bool = false,
            uploadsLiveCameraFrame: Bool = false,
            sendsPrivatePhoto: Bool = false,
            sendsIdentityData: Bool = false,
            sendsPreciseLocation: Bool = false,
            sendsRawLearningEvents: Bool = false,
            allowsGenerativeImageOutput: Bool = false
        ) {
            self.singlePhoneOnly = singlePhoneOnly
            self.usesAuditedPayload = usesAuditedPayload
            self.usesProviderHealthGate = usesProviderHealthGate
            self.storesRawPhoto = storesRawPhoto
            self.uploadsLiveCameraFrame = uploadsLiveCameraFrame
            self.sendsPrivatePhoto = sendsPrivatePhoto
            self.sendsIdentityData = sendsIdentityData
            self.sendsPreciseLocation = sendsPreciseLocation
            self.sendsRawLearningEvents = sendsRawLearningEvents
            self.allowsGenerativeImageOutput = allowsGenerativeImageOutput
        }

        public var isSafeForSinglePhoneCreativeReasoning: Bool {
            singlePhoneOnly
                && usesAuditedPayload
                && usesProviderHealthGate
                && !storesRawPhoto
                && !uploadsLiveCameraFrame
                && !sendsPrivatePhoto
                && !sendsIdentityData
                && !sendsPreciseLocation
                && !sendsRawLearningEvents
                && !allowsGenerativeImageOutput
        }
    }
}

public struct HealthGatedCreativeInterpretationAdapter: Sendable {
    private let provider: any CreativeInterpretationReasoningProvider

    public init(provider: any CreativeInterpretationReasoningProvider = CreativeInterpretationHeuristicProvider()) {
        self.provider = provider
    }

    public func interpret(
        for plan: CreativeInterpretationPlan,
        providerHealthSnapshot: OnlineInspirationHealthSnapshot?,
        maxResponseWords: Int = 120,
        generatedAt: Date = Date()
    ) async throws -> CreativeInterpretationResponse {
        let request = try CreativeInterpretationRequest(
            plan: plan,
            provider: provider.provider,
            maxResponseWords: maxResponseWords
        )
        let healthGate = CreativeInterpretationProviderHealthGate.make(
            for: request,
            healthSnapshot: providerHealthSnapshot
        )

        guard healthGate.canRunProvider else {
            throw CreativeInterpretationAdapterError.blockedByHealthGate(healthGate.deniedReasons)
        }

        let result = try await provider.interpret(request: request)
        guard !result.guidance.isEmpty else {
            throw CreativeInterpretationAdapterError.emptyProviderOutput
        }

        let response = CreativeInterpretationResponse(
            planId: request.planId,
            provider: request.provider,
            result: result,
            maxResponseWords: request.maxResponseWords,
            generatedAt: generatedAt,
            payloadAudit: request.payloadAudit,
            healthGate: healthGate
        )

        guard response.privacy.isSafeForSinglePhoneCreativeReasoning else {
            throw CreativeInterpretationAdapterError.unsafeProviderResponse
        }

        return response
    }
}

File: backend/api/creative-interpretation.mjs
export const lensPilotCreativeInterpretationApiDefaults = {
  apiVersion: "2026-08-31",
  path: "/v1/creative-interpretation",
  method: "POST",
  requiresServerSideOpenAIKey: true,
  acceptsClientOpenAIKey: false,
  openAIEndpoint: "https://api.openai.com/v1/responses",
  openAIModel: "gpt-5.6-luna",
  maxToolCalls: 2,
};

export const lensPilotCreativeInterpretationApiPrivacy = {
  keepsOpenAIKeyOnServer: true,
  acceptsClientOpenAIKey: false,
  singlePhoneOnly: true,
  usesAuditedPayload: true,
  usesProviderHealthGate: true,
  storesRawPhoto: false,
  uploadsLiveCameraFrame: false,
  sendsPrivatePhoto: false,
  sendsIdentityData: false,
  sendsPreciseLocation: false,
  sendsRawLearningEvents: false,
  allowsGenerativeImageOutput: false,
};

const requiredCreativeInterpretationMustNotSendTerms = [
  "raw_live_camera_feed",
  "private_photo",
  "face_identity",
  "precise_location_without_consent",
  "raw_learning_events",
];

const creativeInterpretationBlockedPayloadTerms = [
  "raw_live_camera",
  "private_photo",
  "face_identity",
  "identity_recognition",
  "precise_location",
  "gps",
  "latitude",
  "longitude",
  "exif",
  "raw_learning_event",
  "base64",
  "image_data",
  "photo_bytes",
];

const unsafeCreativeInterpretationProviderOutputTerms = [
  "raw_live_camera",
  "raw camera frame",
  "private_photo",
  "private photo",
  "face_identity",
  "face identity",
  "identity_recognition",
  "identity recognition",
  "precise_location",
  "precise location",
  "gps",
  "latitude",
  "longitude",
  "exif",
  "raw_learning_event",
  "raw learning event",
  "base64",
  "image_data",
  "photo_bytes",
  "generate an image",
  "generative edit",
  "sky replacement",
  "object removal",
];

const openAICreativeInterpretationInstructions =
  "You are LensPilot AI's photography reasoning provider. Return only JSON matching the schema. Use only the audited text summary and public-reference context in the request. Do not ask for or mention uploading live camera frames, private photos, identity data, precise location, raw learning events, EXIF, base64, or photo bytes. Keep guidance capture-realistic, concise, and useful for one phone in the user's hand. Do not promise generated edits, object removal, sky replacement, or a synthetic final image.";

const openAICreativeInterpretationResponseFormat = {
  type: "json_schema",
  name: "lenspilot_creative_interpretation",
  strict: true,
  schema: {
    type: "object",
    additionalProperties: false,
    properties: {
      headline: {
        type: "string",
        maxLength: 96,
      },
      guidance: {
        type: "array",
        minItems: 2,
        maxItems: 4,
        items: {
          type: "string",
          maxLength: 180,
        },
      },
    },
    required: ["headline", "guidance"],
  },
};

export function createLensPilotCreativeInterpretationApi(options = {}) {
  const environment = options.environment ?? globalThis.process?.env ?? {};
  const config = {
    openAIAPIKey: cleanOptional(options.openAIAPIKey ?? environment.OPENAI_API_KEY),
    expectedClientToken: cleanOptional(options.expectedClientToken ?? environment.LENSPILOT_CREATIVE_API_TOKEN),
    openAIEndpoint: cleanOptional(options.openAIEndpoint) ?? lensPilotCreativeInterpretationApiDefaults.openAIEndpoint,
    model: cleanOptional(options.model ?? environment.LENSPILOT_OPENAI_MODEL) ?? lensPilotCreativeInterpretationApiDefaults.openAIModel,
    allowsWebSearch: parseBooleanOption(options.allowsWebSearch ?? environment.LENSPILOT_OPENAI_WEB_SEARCH, true),
    fetchImpl: options.fetchImpl ?? globalThis.fetch,
    now: options.now ?? (() => new Date().toISOString()),
  };

  return {
    async handle(requestLike) {
      return handleCreativeInterpretationRequest(requestLike, config);
    },
  };
}

export function createLensPilotCreativeInterpretationFetchHandler(options = {}) {
  const api = createLensPilotCreativeInterpretationApi(options);
  return async function handleFetchRequest(request) {
    const result = await api.handle(request);
    return new Response(result.body, {
      status: result.status,
      headers: result.headers,
    });
  };
}

export function makeOpenAICreativeInterpretationResponsesPayload(request, options = {}) {
  if (!isCreativeInterpretationRequestSafe(request)) {
    throw new Error("unsafe_openai_creative_interpretation_request");
  }

  const payload = {
    model: cleanOptional(options.model) ?? lensPilotCreativeInterpretationApiDefaults.openAIModel,
    instructions: openAICreativeInterpretationInstructions,
    input: openAICreativeInterpretationInputText(request),
    store: false,
    max_output_tokens: Math.min(640, Math.max(96, request.maxResponseWords * 3)),
    reasoning: {
      effort: "low",
    },
    text: {
      format: openAICreativeInterpretationResponseFormat,
    },
    metadata: {
      lenspilot_plan_id: String(request.planId).slice(0, 64),
      lenspilot_provider: request.provider,
      lenspilot_payload: "audited_text_only",
    },
  };

  if (options.allowsWebSearch ?? true) {
    payload.tools = [{ type: "web_search" }];
    payload.tool_choice = "auto";
    payload.max_tool_calls = Math.min(
      4,
      Math.max(1, Math.trunc(options.maxToolCalls ?? lensPilotCreativeInterpretationApiDefaults.maxToolCalls))
    );
  }

  return payload;
}

export function parseOpenAICreativeInterpretationProviderResult(payload) {
  if (!isRecord(payload)) {
    throw new Error("invalid_openai_creative_interpretation_response");
  }

  if (isRecord(payload.error)) {
    throw new Error("openai_creative_interpretation_api_error");
  }

  if (typeof payload.status === "string" && payload.status !== "completed") {
    throw new Error(`openai_creative_interpretation_incomplete:${payload.status}`);
  }

  const outputText = typeof payload.output_text === "string"
    ? payload.output_text
    : firstOpenAIOutputText(payload.output);
  if (!outputText) {
    throw new Error("missing_openai_creative_interpretation_output_text");
  }

  let parsed;
  try {
    parsed = JSON.parse(outputText);
  } catch {
    throw new Error("invalid_openai_creative_interpretation_json");
  }

  if (!isRecord(parsed) || typeof parsed.headline !== "string" || !Array.isArray(parsed.guidance)) {
    throw new Error("invalid_openai_creative_interpretation_json");
  }

  const result = makeCreativeInterpretationProviderResult(
    parsed.headline,
    parsed.guidance.filter((item) => typeof item === "string")
  );
  if (!isCreativeInterpretationProviderResultSafe(result)) {
    throw new Error("unsafe_creative_interpretation_provider_output");
  }

  return result;
}

export function makeLensPilotCreativeInterpretationApiResponse(apiRequest, result, generatedAt) {
  if (!isLensPilotCreativeInterpretationApiRequestSafe(apiRequest)) {
    throw new Error("unsafe_lenspilot_creative_interpretation_api_request");
  }
  if (!isCreativeInterpretationProviderResultSafe(result) || result.guidance.length === 0) {
    throw new Error("unsafe_lenspilot_creative_interpretation_api_response");
  }

  return {
    apiVersion: lensPilotCreativeInterpretationApiDefaults.apiVersion,
    status: "completed",
    provider: apiRequest.request.provider,
    result,
    generatedAt,
    privacy: lensPilotCreativeInterpretationApiPrivacy,
  };
}

export function isLensPilotCreativeInterpretationApiRequestSafe(apiRequest) {
  return isRecord(apiRequest) &&
    apiRequest.apiVersion === lensPilotCreativeInterpretationApiDefaults.apiVersion &&
    isRecord(apiRequest.client) &&
    apiRequest.client.platform === "ios" &&
    isCreativeInterpretationRequestSafe(apiRequest.request) &&
    isCreativeInterpretationProviderHealthGateSafe(apiRequest.healthGate) &&
    !containsClientOpenAIKey(apiRequest);
}

async function handleCreativeInterpretationRequest(requestLike, config) {
  const method = String(requestLike?.method ?? "GET").toUpperCase();
  if (method !== lensPilotCreativeInterpretationApiDefaults.method) {
    return jsonResponse(405, {
      error: {
        code: "method_not_allowed",
        message: "Use POST for creative interpretation.",
      },
    });
  }

  const headers = normalizeHeaders(requestLike?.headers);
  if (containsClientOpenAIKey(headers)) {
    return jsonResponse(400, {
      error: {
        code: "client_openai_key_rejected",
        message: "OpenAI keys must stay on the server.",
      },
    });
  }

  if (config.expectedClientToken) {
    const expectedAuthorization = `Bearer ${config.expectedClientToken}`;
    if (headerValue(headers, "authorization") !== expectedAuthorization) {
      return jsonResponse(401, {
        error: {
          code: "unauthorized",
          message: "Creative API client authorization failed.",
        },
      });
    }
  }

  let body;
  try {
    body = await parseJSONBody(requestLike);
  } catch {
    return jsonResponse(400, {
      error: {
        code: "invalid_json_body",
        message: "Creative API request body must be valid JSON.",
      },
    });
  }

  if (!isRecord(body) || containsClientOpenAIKey(body)) {
    return jsonResponse(400, {
      error: {
        code: "unsafe_request",
        message: "Creative API request body is unsafe.",
      },
    });
  }

  const apiRequest = {
    apiVersion: body.apiVersion,
    request: body.request,
    healthGate: body.healthGate,
    client: body.client,
  };
  if (!isLensPilotCreativeInterpretationApiRequestSafe(apiRequest)) {
    return jsonResponse(400, {
      error: {
        code: "unsafe_request",
        message: "Creative API request failed the single-phone safety gate.",
      },
    });
  }

  if (!config.openAIAPIKey) {
    return jsonResponse(500, {
      error: {
        code: "missing_server_openai_key",
        message: "Creative API is not configured.",
      },
    });
  }

  if (typeof config.fetchImpl !== "function") {
    return jsonResponse(500, {
      error: {
        code: "fetch_unavailable",
        message: "Creative API fetch runtime is unavailable.",
      },
    });
  }

  const openAIPayload = makeOpenAICreativeInterpretationResponsesPayload(apiRequest.request, {
    model: config.model,
    allowsWebSearch: config.allowsWebSearch,
    maxToolCalls: lensPilotCreativeInterpretationApiDefaults.maxToolCalls,
  });

  let openAIResponse;
  try {
    openAIResponse = await config.fetchImpl(config.openAIEndpoint, {
      method: "POST",
      headers: {
        "Authorization": `Bearer ${config.openAIAPIKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(openAIPayload),
    });
  } catch {
    return jsonResponse(502, {
      error: {
        code: "openai_request_failed",
        message: "Creative provider request failed.",
      },
    });
  }

  if (!openAIResponse?.ok) {
    return jsonResponse(502, {
      error: {
        code: "openai_request_failed",
        message: "Creative provider returned an unsuccessful status.",
      },
    });
  }

  let openAIJSON;
  try {
    openAIJSON = await readJSONResponse(openAIResponse);
  } catch {
    return jsonResponse(502, {
      error: {
        code: "openai_invalid_json",
        message: "Creative provider response could not be read.",
      },
    });
  }

  let result;
  try {
    result = parseOpenAICreativeInterpretationProviderResult(openAIJSON);
  } catch {
    return jsonResponse(502, {
      error: {
        code: "unsafe_provider_response",
        message: "Creative provider response failed safety validation.",
      },
    });
  }

  return jsonResponse(200, makeLensPilotCreativeInterpretationApiResponse(apiRequest, result, config.now()));
}

function isCreativeInterpretationRequestSafe(request) {
  return isRecord(request) &&
    request.provider === "online_reasoning" &&
    isRecord(request.payloadAudit) &&
    request.payloadAudit.safeToSend === true &&
    Array.isArray(request.payloadAudit.deniedReasons) &&
    request.payloadAudit.deniedReasons.length === 0 &&
    Array.isArray(request.payloadAudit.blockedTermsDetected) &&
    request.payloadAudit.blockedTermsDetected.length === 0 &&
    isNonEmptyStringArray(request.allowedInputs) &&
    isNonEmptyStringArray(request.mustNotSend) &&
    isNonEmptyStringArray(request.inputSummary) &&
    isNonEmptyStringArray(request.suggestionBriefs) &&
    Number.isFinite(request.maxResponseWords) &&
    request.maxResponseWords >= 40 &&
    request.maxResponseWords <= 240 &&
    isCreativeInterpretationPrivacySafe(request.privacy) &&
    requiredCreativeInterpretationMustNotSendTerms.every((term) => request.mustNotSend.includes(term)) &&
    detectedCreativeInterpretationBlockedPayloadTerms(request).length === 0;
}

function isCreativeInterpretationPrivacySafe(privacy) {
  return isRecord(privacy) &&
    privacy.singlePhoneOnly === true &&
    privacy.requiresUserConsent === true &&
    privacy.sendsRawCameraFrame === false &&
    privacy.sendsPrivatePhoto === false &&
    privacy.sendsIdentityData === false &&
    privacy.sendsPreciseLocation === false &&
    privacy.sendsRawLearningEvents === false &&
    privacy.allowsGenerativeOutput === false;
}

function isCreativeInterpretationProviderHealthGateSafe(healthGate) {
  return isRecord(healthGate) &&
    healthGate.canRunProvider === true &&
    (healthGate.providerHealthStatus === "available" || healthGate.providerHealthStatus === "degraded") &&
    Number.isFinite(healthGate.publicReferenceCount) &&
    healthGate.publicReferenceCount > 0 &&
    isRecord(healthGate.payloadAudit) &&
    healthGate.payloadAudit.safeToSend === true &&
    Array.isArray(healthGate.payloadAudit.deniedReasons) &&
    healthGate.payloadAudit.deniedReasons.length === 0 &&
    Array.isArray(healthGate.payloadAudit.blockedTermsDetected) &&
    healthGate.payloadAudit.blockedTermsDetected.length === 0 &&
    isRecord(healthGate.privacy) &&
    healthGate.privacy.singlePhoneOnly === true &&
    healthGate.privacy.requiresUserConsent === true &&
    healthGate.privacy.sendsRawCameraFrame === false &&
    healthGate.privacy.sendsPrivatePhoto === false &&
    healthGate.privacy.sendsIdentityData === false &&
    healthGate.privacy.sendsPreciseLocation === false &&
    healthGate.privacy.sendsRawLearningEvents === false;
}

function detectedCreativeInterpretationBlockedPayloadTerms(request) {
  const inspectedText = [
    ...(Array.isArray(request.inputSummary) ? request.inputSummary : []),
    ...(Array.isArray(request.suggestionBriefs) ? request.suggestionBriefs : []),
  ].join(" ").toLowerCase();
  return creativeInterpretationBlockedPayloadTerms.filter((term) => inspectedText.includes(term));
}

function openAICreativeInterpretationInputText(request) {
  const summary = request.inputSummary
    .slice(0, 8)
    .map((item) => `- ${item}`)
    .join("\n");
  const suggestions = request.suggestionBriefs
    .slice(0, 6)
    .map((item) => `- ${item}`)
    .join("\n");

  return [
    "LensPilot creative interpretation request.",
    `Plan id: ${request.planId}`,
    `Allowed input classes: ${request.allowedInputs.join(", ")}`,
    `Max response words: ${request.maxResponseWords}`,
    "",
    "Safe input summary:",
    summary,
    "",
    "Candidate capture guidance:",
    suggestions,
    "",
    "Return a short headline and 2-4 capture-realistic guidance strings.",
  ].join("\n");
}

function makeCreativeInterpretationProviderResult(headline, guidance) {
  return {
    headline: cleanProviderText(headline, 96),
    guidance: guidance
      .map((item) => cleanProviderText(item, 180))
      .filter(Boolean)
      .slice(0, 4),
  };
}

function isCreativeInterpretationProviderResultSafe(result) {
  const inspectedText = [result.headline, ...result.guidance].join(" ").toLowerCase();
  return !unsafeCreativeInterpretationProviderOutputTerms.some((term) => inspectedText.includes(term));
}

function firstOpenAIOutputText(output) {
  if (!Array.isArray(output)) return undefined;

  for (const item of output) {
    if (!isRecord(item) || !Array.isArray(item.content)) continue;

    for (const content of item.content) {
      if (isRecord(content) && content.type === "output_text" && typeof content.text === "string") {
        return content.text;
      }
    }
  }

  return undefined;
}

async function parseJSONBody(requestLike) {
  if (typeof requestLike?.json === "function") {
    return requestLike.json();
  }

  const body = requestLike?.body;
  if (typeof body === "string") {
    return JSON.parse(body);
  }
  if (body instanceof Uint8Array) {
    return JSON.parse(new TextDecoder().decode(body));
  }
  if (isRecord(body)) {
    return body;
  }

  throw new Error("invalid_json_body");
}

async function readJSONResponse(response) {
  if (typeof response.json === "function") {
    return response.json();
  }
  if (typeof response.text === "function") {
    return JSON.parse(await response.text());
  }
  if (typeof response.body === "string") {
    return JSON.parse(response.body);
  }
  if (isRecord(response.body)) {
    return response.body;
  }

  throw new Error("invalid_json_response");
}

function normalizeHeaders(headers) {
  if (!headers) return {};
  if (typeof headers.get === "function") {
    const normalized = {};
    for (const key of ["authorization", "content-type", "x-lenspilot-client"]) {
      const value = headers.get(key);
      if (value) normalized[key] = value;
    }
    return normalized;
  }
  if (Array.isArray(headers)) {
    return Object.fromEntries(headers.map(([key, value]) => [String(key).toLowerCase(), String(value)]));
  }
  if (isRecord(headers)) {
    return Object.fromEntries(
      Object.entries(headers).map(([key, value]) => [key.toLowerCase(), String(value)])
    );
  }
  return {};
}

function headerValue(headers, name) {
  return headers[name.toLowerCase()];
}

function containsClientOpenAIKey(value) {
  if (typeof value === "string") {
    const normalized = value.toLowerCase();
    return normalized.includes("openai_api_key") || /(^|[\s"'=:])sk-(proj-)?[a-z0-9_-]{8,}/i.test(value);
  }

  if (Array.isArray(value)) {
    return value.some(containsClientOpenAIKey);
  }

  if (isRecord(value)) {
    return Object.entries(value).some(([key, nestedValue]) => {
      const normalizedKey = key.toLowerCase();
      return normalizedKey === "openai_api_key" ||
        normalizedKey === "apikey" ||
        normalizedKey === "api_key" ||
        containsClientOpenAIKey(nestedValue);
    });
  }

  return false;
}

function parseBooleanOption(value, fallback) {
  const cleaned = cleanOptional(value);
  if (!cleaned) return fallback;
  return !["0", "false", "no", "off"].includes(cleaned.toLowerCase());
}

function cleanOptional(value) {
  if (typeof value !== "string") return undefined;
  const trimmed = value.trim();
  return trimmed.length > 0 ? trimmed : undefined;
}

function cleanProviderText(value, maxLength) {
  const collapsed = value
    .split(/\s+/)
    .join(" ")
    .trim();
  return collapsed.length <= maxLength ? collapsed : collapsed.slice(0, maxLength).trim();
}

function jsonResponse(status, payload) {
  return {
    status,
    headers: {
      "content-type": "application/json; charset=utf-8",
      "cache-control": "no-store",
    },
    body: JSON.stringify(payload),
  };
}

function isRecord(value) {
  return typeof value === "object" && value !== null;
}

function isNonEmptyStringArray(value) {
  return Array.isArray(value) &&
    value.length > 0 &&
    value.every((item) => typeof item === "string" && item.trim().length > 0);
}
